eventsourcing/src/main/java/org/axonframework/eventsourcing/eventstore/ContinuousMessageStream.java

```diff
@@ -0,0 +1,181 @@
+    private Entry<EventMessage> nextEntry;
+    private Throwable error;
+    private int position;  // position within data
+    private Registration callbackRegistration;
+    private Runnable callback;
+    private boolean closed;
+    private boolean completed;
+
+    /**
+     * Creates a new {@code ContinuousMessageStream} instance configured with the given strategies.
+     *
+     * @param fetcher          a function that, given the last fetched element (or {@code null} for the first call),
+     *                         retrieves the next batch of elements; must not return {@code null}
+     * @param converter        a function converting each fetched element into an {@link Entry} containing an {@link EventMessage}
+     * @param callbackTracker  a function that, given this stream and a callback {@link Runnable}, registers
+     *                         a listener and returns a {@link Registration} allowing it to be canceled
+     */
+    public ContinuousMessageStream(
+        Function<E, List<E>> fetcher,
+        Function<E, Entry<EventMessage>> converter,
+        BiFunction<ContinuousMessageStream<?>, Runnable, Registration> callbackTracker
+    ) {
+        this.fetcher = fetcher;
+        this.converter = converter;
+        this.callbackTracker = callbackTracker;
+    }
+
+    @Override
+    public synchronized void setCallback(Runnable callback) {
+        if (!closed) {
+            this.callback = callback;
+            this.callbackRegistration = callbackTracker.apply(this, this::invokeCallback);
+
+            invokeCallback();
+        }
+    }
+
+    @Override
+    public synchronized Optional<Entry<EventMessage>> next() {
+        try {
+            return peek();
+        }
+        finally {
+            nextEntry = null;
+        }
+    }
+
+    @Override
+    public synchronized Optional<Entry<EventMessage>> peek() {
+        if (closed) {
+            return Optional.empty();
+        }
+
+        if (nextEntry == null) {
+            if (position >= data.size()) {
+                fetchMore();  // TODO may block, this is against the contract of peek/next/hasNextAvailable
+
+                if (data.isEmpty()) {
+                    return Optional.empty();
+                }
+            }
+
+            E element = data.get(position++);
+
+            nextEntry = converter.apply(element);
+        }
+
+        return Optional.of(nextEntry);
+    }
+
+    @Override
+    public synchronized Optional<Throwable> error() {
+        return Optional.ofNullable(error);
+    }
+
+    @Override
+    public synchronized boolean isCompleted() {
+        return completed;
+    }
+
+    @Override
+    public synchronized boolean hasNextAvailable() {
+        return peek().isPresent();
+    }
+
+    @Override
+    public synchronized void close() {
+        if (!closed) {
+            closed = true;
+            completed = true;  // the only way to complete an infinite stream is to close it
```

Comment:
We should not mark the stream as completed, since we decided not to read till the end. Formally, the stream is still incomplete.